### ISSUE TITLE
Question: how to use ctags to generate multiple entries for name-equal local variables in the same file?

### ISSUE THREAD

**Issue by sshsu** [universal-ctags/ctags#1956]
I want to use ctags to make tags for the file test.cpp

```test.cpp
#inlcude<iostream>

int fn1(){
   int temp;
}

int fn2(){
   int temp;
   temp = 0;
}

int main(){

}
```
I use the next command for generating tags file

`ctags --c++-kind=+lf test.cpp`
tags file content shown as below

```
!_TAG_FILE_FORMAT       2       /extended format; --format=1 will not append ;" to lines/
!_TAG_FILE_SORTED       1       /0=unsorted, 1=sorted, 2=foldcase/
!_TAG_PROGRAM_AUTHOR    Darren Hiebert  /dhiebert@users.sourceforge.net/
!_TAG_PROGRAM_NAME      Exuberant Ctags //
!_TAG_PROGRAM_URL       http://ctags.sourceforge.net    /official site/
!_TAG_PROGRAM_VERSION   5.9~svn20110310 //
fn      test.cpp        /^int fn(){$/;" f
fn1     test.cpp        /^int fn1(){$/;"        f
main    test.cpp        /^int main (){$/;"      f
temp    test.cpp        /^    int temp;$/;"     l
```
when I use shortcut "g]" at fn2() temp in vim with the aim of jumping to defination of variable temp in fn2() ,then the menu appears and there are only one entry shown as below

```
  # pri kind tag               file
  1 F C l    temp              test.cpp
               int temp;
  Type number and <Enter> (empty cancels): 
```
I input the number "1" , it jump to the funciton fn1()'s defination of variable temp instead of fn2()'s,

and right now I using shortcut "gd" at fn2()'s temp, it also jumps to fn1()'s, really troubles me.

how should I do for generating different tags entries for different local vairables which support to jump to different positions?

It seems that option "--fields=+n" doesn't work

thanks

**Comment by masatake** [comment]
```
!_TAG_PROGRAM_NAME      Exuberant Ctags //
```

We maintain Universal-ctags, unofficial fork of Exuberant Ctags.
If you have a trouble with  Exuberant Ctags, you can report it to https://sourceforge.net/projects/ctags/.
However, the project is not active.

u-ctags generates tags that includes both temp of fn1 and temp of fn2.
u-ctags prints scope information by default. So you can distinguish them.
Using "--fields=+n" looks good idea.
"-n' can be used, too.
 
```
[yamato@slave]/tmp% cat foo.cpp                                
#inlcude<iostream>

int fn1(){
   int temp;
}

int fn2(){
   int temp;
   temp = 0;
}

int main(){

}
[yamato@slave]/tmp% u-ctags --kinds-c++=+lf -o - foo.cpp            
fn1	foo.cpp	/^int fn1(){$/;"	f	typeref:typename:int
fn2	foo.cpp	/^int fn2(){$/;"	f	typeref:typename:int
main	foo.cpp	/^int main(){$/;"	f	typeref:typename:int
temp	foo.cpp	/^   int temp;$/;"	l	function:fn1	typeref:typename:int	file:
temp	foo.cpp	/^   int temp;$/;"	l	function:fn2	typeref:typename:int	file:
[yamato@slave]/tmp% u-ctags -n --kinds-c++=+lf -o - foo.cpp
fn1	foo.cpp	3;"	f	typeref:typename:int
fn2	foo.cpp	7;"	f	typeref:typename:int
main	foo.cpp	12;"	f	typeref:typename:int
temp	foo.cpp	4;"	l	function:fn1	typeref:typename:int	file:
temp	foo.cpp	8;"	l	function:fn2	typeref:typename:int	file:
[yamato@slave]/tmp% u-ctags  --kinds-c++=+lf --fields=+n -o - foo.cpp
fn1	foo.cpp	/^int fn1(){$/;"	f	line:3	typeref:typename:int
fn2	foo.cpp	/^int fn2(){$/;"	f	line:7	typeref:typename:int
main	foo.cpp	/^int main(){$/;"	f	line:12	typeref:typename:int
temp	foo.cpp	/^   int temp;$/;"	l	line:4	function:fn1	typeref:typename:int	file:
temp	foo.cpp	/^   int temp;$/;"	l	line:8	function:fn2	typeref:typename:int	file:
```

As far as I tried these option with Exuberant ctags, both temp local variables are printed.
You wrote:
> It seems that option "--fields=+n" doesn't work
I guess this is not an issue of ctags side.
I think it is an issue of your editor, maybe vim.

**Comment by sshsu** [comment]
@masatake, many thanks for your assistance, I don't realize there are two different verions ctags until now, and I have changed my e-ctags to u-ctags, follow your adivce, I try it again, but still some problems,

```
root@iZwz957qhjacaodhugl3a7Z:~/project/test# cat test.cpp
#inlcude<iostream>

int fn1(){
   int temp;
}

int fn2(){
   int temp;
   temp = 0;
}

int main(){

}
root@iZwz957qhjacaodhugl3a7Z:~/project/test# ctags  --kinds-c++=+lf --fields=+n test.cpp
root@iZwz957qhjacaodhugl3a7Z:~/project/test# cat tags
!_TAG_FILE_FORMAT       2       /extended format; --format=1 will not append ;" to lines/
!_TAG_FILE_SORTED       1       /0=unsorted, 1=sorted, 2=foldcase/
!_TAG_OUTPUT_MODE       u-ctags /u-ctags or e-ctags/
!_TAG_PROGRAM_AUTHOR    Universal Ctags Team    //
!_TAG_PROGRAM_NAME      Universal Ctags /Derived from Exuberant Ctags/
!_TAG_PROGRAM_URL       https://ctags.io/       /official site/
!_TAG_PROGRAM_VERSION   0.0.0   /5b42263/
fn1     test.cpp        /^int fn1(){$/;"        f       line:3  typeref:typename:int
fn2     test.cpp        /^int fn2(){$/;"        f       line:7  typeref:typename:int
main    test.cpp        /^int main(){$/;"       f       line:12 typeref:typename:int
temp    test.cpp        /^   int temp;$/;"      l       line:4  function:fn1    typeref:typename:int    file:
temp    test.cpp        /^   int temp;$/;"      l       line:8  function:fn2    typeref:typename:int    file:
```
I use "vim test.cpp" to edit,  move the cursor at the fn2' temp as below 
```
int fn2(){
   int temp;
   temp = 0;  // move the cursor at this "temp" 
}
```
then use shortcut "g]"  intending to make a optional jump, jump menu shows as expect
![screen](https://user-images.githubusercontent.com/26362180/50195410-e63a0900-0379-11e9-9021-1f20d43c8734.jpg)
However, when I input 2 and enter it, the cursor jumps to fn1() temp（use shortcut "ctrl ]" or "gd" at fn2' temp also the same)
```
int fn1(){
     int temp;  //jump to here
}
```
That 's where I said options "--fields=+n" doesn't work. Though it makes more than one entry for name equal local variables , line and scope information in the tags entry don't make effect when vim use the tags entry to match and jump，vim just uses regexp to match and jump
```
temp    test.cpp        /^   int temp;$/;"      l       line:8  function:fn2    typeref:typename:int    file:
```
and I have found some people have mentiond [this] (https://github.com/vim/vim/issues/1057) in 2016，I haven't made sure whether vim supports this and haven't found out a good and simple enghou solution yet，welcome to give a good solution

in addition, u-ctags -n works good at jumping different positions  for the sake that the symbol position in  tag file  is line number instead of regexp, as below 
```
[yamato@slave]/tmp% u-ctags -n --kinds-c++=+lf -o - foo.cpp
fn1	foo.cpp	3;"	f	typeref:typename:int
```

**Comment by masatake** [comment]
O.k. I will implement what I proposed in #1125.

**Comment by masatake** [comment]
Can I ask you to review man page when I implemented `--excmd=both` ?

**Comment by sshsu** [comment]
> See and try -excmd=combined introduced in #1959.
> How do you think the name "combined"? "combine" is better?

OK, I am trying it now , I think "combined" is better, consistent with "mixed"

**Comment by sshsu** [comment]
sorry , I see the man page is --excmd="mix', so I think combine might be more suitable.
And I am not familiar with github, so trying so far but don't know how to pull you new commit to test.  Any tutorial?  or,  I can't pull your new commit until your pull request merge?

**Comment by masatake** [comment]
> sorry , I see the man page is --excmd="mix', so I think combine might be more suitable.
I see. Thank you.

> And I am not familiar with github, so trying so far but don't know how to pull you new commit to test. Any tutorial? or, I can't pull your new commit until your pull request merge?

1.  do git clone https://github.com/universal-ctags/ctags.git if you have not done yet
2. do cd ctags
3. do git checkout -b masatake-excmd-combined master
4. do git pull https://github.com/masatake/ctags.git excmd-combined
5. bash ./autogen.sh
6. ./configure; make
7. run ./ctags with --excmd=combined

**Comment by sshsu** [comment]
so nice of you, all it works, thanks

**Comment by masatake** [comment]
Thank you for testing. After changing the name from combined to combine, I will merge this change.

**Comment by masatake** [comment]
As far as reading https://github.com/vim/vim/pull/2546, https://github.com/vim/vim/issues/1057, and  https://github.com/vim/vim/issues/1057, I should not use line number directly. Instead, we should use line number - 1. How do you think?

**Comment by sshsu** [comment]
> As far as reading [vim/vim#2546](https://github.com/vim/vim/pull/2546), #1599, and [vim/vim#1057](https://github.com/vim/vim/issues/1057), I should not use line number directly. Instead, we should use line number - 1. How do you think?

sorry for late,  you are right,  use regex to match and search,  the cursor will  jump to matached position starting from current line's next line. 

use "line number -1" is good ,that can fix it. Is the line number will never be zero, right?  If no, "lin number -1" will be -1 when the value of line numer is zero,  in vim, jump to -1 means jumping to the previous line, it might cause a little problem in this case.

**Comment by sshsu** [comment]
and one more thing, for a test file as below
```
int temp;
int temp;

temp; //use g + ctrl + ] here
```

to generate tag file and mannully minus -1 of lthe value of line number,
```
!_TAG_FILE_FORMAT       2       /extended format; --format=1 will not append ;" to lines/
!_TAG_FILE_SORTED       1       /0=unsorted, 1=sorted, 2=foldcase/
!_TAG_OUTPUT_MODE       u-ctags /u-ctags or e-ctags/
!_TAG_PROGRAM_AUTHOR    Universal Ctags Team    //
!_TAG_PROGRAM_NAME      Universal Ctags /Derived from Exuberant Ctags/
!_TAG_PROGRAM_URL       https://ctags.io/       /official site/
!_TAG_PROGRAM_VERSION   0.0.0   /a9b1006/
temp    main.cpp        0;/^int temp;$/;"       v       typeref:typename:int
temp    main.cpp        1;/^int temp;$/;"       v       typeref:typename:int
```
the first variable temp's line number is 0, and use this tags entry will jump to the first temp as we wish.

directly use that in vim also work
```
:0;/^int temp;$/;
```

however , When I do it separately as below，it jumps to the second one, amazing
```
:0
:/^int temp;$/;
```

**Comment by masatake** [comment]
I see. Thank you. Maybe what you find at the end is about an issue of vim side.

**Comment by masatake** [comment]
Solved via #1959.
Use --excmd=combine option.
